django/thunderstore/social/providers.py

```diff
@@ -0,0 +1,112 @@
+from typing import Dict
+
+import requests
+from django.conf import settings
+
+
+class BaseOauthHelper:
+    token = None
+
+    def __init__(self, code: str, redirect_url: str) -> None:
+        self.code = code
+        self.redirect_url = redirect_url
+
+    def complete_login(self) -> None:
+        """
+        Fetch access token with the temporary code and configured secret.
+        """
+        data = {
+            "client_id": self.CLIENT_ID,
+            "client_secret": self.CLIENT_SECRET,
+            "code": self.code,
+            "grant_type": "authorization_code",  # Required by Discord
+            "redirect_uri": self.redirect_url,  # Required by Discord
+        }
+        headers = {"Accept": "application/json"}
+
+        response = requests.post(self.OAUTH_URL, data, headers=headers)
+        response_data = response.json()
+        self.token = response_data["access_token"]
+
+    def get_user_info(self) -> Dict:
+        raise NotImplementedError()
+
+    def _fetch_from_api(self, url: str) -> Dict:
+        """
+        Use access token to fetch info from provider's public API.
+        """
+        if self.token is None:
+            raise Exception("No token found. Did you call .complete_login()?")
+
+        headers = {"Authorization": f"{self.AUTH_HEADER_KEYWORD} {self.token}"}
+
+        response = requests.get(url, headers=headers)
+        return response.json()
```

Comment:
Instead of returning an arbitrary blob of data, it might be a good idea to add provider-specific validation. I'd suggest using [pydantic](https://pydantic-docs.helpmanual.io/) for that as it's super straight forward (just declare a class with proper type definitions for the data we're interested in & load it). Should allow strong type hinting of the data in methods & proper IDE & CI checks / autocompletes.

This will mean the base class needs to be a generic, which is entirely doable but might get slightly messy. I have some sample code somewhere that follows this paradigm which I can link if you feel it'd be useful.

Reply:
Implemented in a separate PR: https://github.com/thunderstore-io/Thunderstore/pull/577